src/ship_cmd.cpp

```diff
@@ -442,11 +456,25 @@ static Track ChooseShipTrack(Ship *v, TileIndex tile, DiagDirection enterdir, Tr
 
 	bool path_found = true;
 	Track track;
-	switch (_settings_game.pf.pathfinder_for_ships) {
-		case VPF_OPF: track = OPFShipChooseTrack(v, tile, enterdir, tracks, path_found); break;
-		case VPF_NPF: track = NPFShipChooseTrack(v, tile, enterdir, tracks, path_found); break;
-		case VPF_YAPF: track = YapfShipChooseTrack(v, tile, enterdir, tracks, path_found); break;
-		default: NOT_REACHED();
+
+	if (v->dest_tile == 0 || DistanceManhattan(tile, v->dest_tile) > SHIP_MAX_ORDER_DISTANCE + 5) {
+		/* No destination or destination too far, don't invoke pathfinder. */
+		track = TrackBitsToTrack(v->state);
+		if (track != TRACK_X && track != TRACK_Y) track = TrackToOppositeTrack(track);
+		if (!HasBit(tracks, track)) {
+			/* Can't continue in same direction so pick a random available track. */
+			do {
+				track = (Track)RandomRange(TRACK_END);
```

Comment:
I think the code should be rewritten to ensure RandomRange is called just one time.

It works for sure with the current proposal, but a randomiser doesn't feel good here. It seems just a workaround instead of stopping the ship (in the sense of being stuck) and marking it as lost or finding which tracks will lead closer to its destination.

Reply:
Well, there's always FindFirstBit